src/net_processing.cpp

```diff
@@ -1627,41 +1649,7 @@ bool static ProcessHeadersMessage(CNode *pfrom, CConnman *connman, const std::ve
     CBlockHeader first_invalid_header;
     if (!ProcessNewBlockHeaders(headers, state, chainparams, &pindexLast, &first_invalid_header)) {
         if (state.IsInvalid()) {
-            if (punish_duplicate_invalid && state.GetReason() == ValidationInvalidReason::CACHED_INVALID) {
-                // Goal: don't allow outbound peers to use up our outbound
```

Comment:
I think this PR is basically what that TODO is referring to!

Reply:
Also I think most of the information in this comment is needed because the logic is much more complex before this PR than after.  After this PR, it's very easy to see how we handle different kinds of invalid blocks/headers, and so less verbose explanations are needed about the reasoning.  I shrunk this comment down to:
```
            // Disconnect outbound (but not inbound) peers if on an invalid chain.
```
which I think captures all that we're trying to do.